File: apps/web/src/lib/throttle-countdown.test.ts
import { describe, expect, it } from "vitest";
import { formatThrottleMessage, pad2 } from "./throttle-countdown";

describe("throttle-countdown (D-005 — no lockout language, ux-wireframe §3.2)", () => {
  it("pad2 zero-pads single digits", () => {
    expect(pad2(5)).toBe("05");
    expect(pad2(0)).toBe("00");
    expect(pad2(42)).toBe("42");
  });

  it("uses the short-form (seconds) copy at exactly 60s", () => {
    const msg = formatThrottleMessage(60);
    expect(msg).toBe("รอสักครู่แล้วลองใหม่ · เหลือ 60 วินาที");
  });

  it("uses the short-form copy under 60s", () => {
    const msg = formatThrottleMessage(45);
    expect(msg).toBe("รอสักครู่แล้วลองใหม่ · เหลือ 45 วินาที");
  });

  it("uses the long-form (mm:ss) copy above 60s", () => {
    const msg = formatThrottleMessage(107); // 1:47
    expect(msg).toBe("ลองเข้าสู่ระบบถี่เกินไป กรุณารอสักครู่แล้วลองใหม่ · เหลือ 01:47 นาที");
  });

  it("pads minutes and seconds to 2 digits in long form", () => {
    const msg = formatThrottleMessage(3661); // 61:01
    expect(msg).toBe("ลองเข้าสู่ระบบถี่เกินไป กรุณารอสักครู่แล้วลองใหม่ · เหลือ 61:01 นาที");
  });

  it("clamps negative input to 0 (never shows negative countdown)", () => {
    const msg = formatThrottleMessage(-5);
    expect(msg).toBe("รอสักครู่แล้วลองใหม่ · เหลือ 0 วินาที");
  });

  it("rounds fractional seconds up (ceil) so the display never undercounts", () => {
    const msg = formatThrottleMessage(44.2);
    expect(msg).toBe("รอสักครู่แล้วลองใหม่ · เหลือ 45 วินาที");
  });

  it("NEVER contains forbidden lockout language (D-005 hard rule)", () => {
    const forbidden = ["ล็อก", "ระงับ", "แบน"];
    for (const seconds of [5, 45, 60, 61, 107, 3600]) {
      const msg = formatThrottleMessage(seconds);
      for (const word of forbidden) {
        expect(msg).not.toContain(word);
      }
    }
  });
});

File: apps/mobile/test/features/auth/domain/throttle_countdown_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile/features/auth/domain/throttle_countdown.dart';

/// R4 — `formatThrottleMessage` no longer reads copy from `AuthTh` itself
/// (domain purity — see the function's doc comment); this test supplies the
/// SAME literal Thai copy `AuthTh.throttleBannerLong`/`throttleBannerShort`
/// used to hard-code, mirroring exactly what
/// `lib/l10n/app_th.arb`'s `authThrottleBannerLong`/`authThrottleBannerShort`
/// now hold, so every expected string below is unchanged byte-for-byte.
String _format(num remainingSeconds) => formatThrottleMessage(
      remainingSeconds,
      longForm: (mm, ss) => 'ลองเข้าสู่ระบบถี่เกินไป กรุณารอสักครู่แล้วลองใหม่ · เหลือ $mm:$ss นาที',
      shortForm: (n) => 'รอสักครู่แล้วลองใหม่ · เหลือ $n วินาที',
    );

void main() {
  group('pad2', () {
    test('pads single digits', () {
      expect(pad2(5), '05');
    });

    test('leaves two digits unchanged', () {
      expect(pad2(42), '42');
    });
  });

  group('formatThrottleMessage', () {
    test('long form (>60s) uses mm:ss copy', () {
      // 125s -> 02:05
      expect(_format(125), 'ลองเข้าสู่ระบบถี่เกินไป กรุณารอสักครู่แล้วลองใหม่ · เหลือ 02:05 นาที');
    });

    test('boundary exactly 60s uses short form (not long)', () {
      expect(_format(60), 'รอสักครู่แล้วลองใหม่ · เหลือ 60 วินาที');
    });

    test('boundary 61s uses long form', () {
      expect(_format(61), 'ลองเข้าสู่ระบบถี่เกินไป กรุณารอสักครู่แล้วลองใหม่ · เหลือ 01:01 นาที');
    });

    test('short form (<=60s)', () {
      expect(_format(15), 'รอสักครู่แล้วลองใหม่ · เหลือ 15 วินาที');
    });

    test('never never renders "ล็อก"/"ระงับ" (D-005)', () {
      final long = _format(120);
      final short = _format(30);
      expect(long.contains('ล็อก'), isFalse);
      expect(long.contains('ระงับ'), isFalse);
      expect(short.contains('ล็อก'), isFalse);
      expect(short.contains('ระงับ'), isFalse);
    });

    test('clamps negative remaining seconds to 0', () {
      expect(_format(-5), 'รอสักครู่แล้วลองใหม่ · เหลือ 0 วินาที');
    });

    test('rounds up fractional seconds (ceil)', () {
      expect(_format(14.2), 'รอสักครู่แล้วลองใหม่ · เหลือ 15 วินาที');
    });
  });
}
